### PR TITLE
lua: Move utility functions into a separate module

### PR DESCRIPTION
**Pull request by tarruda**
Extracting some lua filter plugin functions into a separate module, where we can more easily reuse in future lua related plugins

I also added a commit which enables cmake to generate clang compilation database (useful for editors/IDEs that support it)

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [N/A] Example configuration file for the change
- [N/A] Debug log output from testing the change

- [x] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

**Documentation**

- [N/A] Documentation required for this feature

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

**Comment by tarruda** [comment]
Valgrind logs for runtime test:
```
$ valgrind --leak-check=full ./build/bin/flb-rt-filter_lua
==46720== Memcheck, a memory error detector
==46720== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==46720== Using Valgrind-3.16.1 and LibVEX; rerun with -h for copyright info
==46720== Command: ./build/bin/flb-rt-filter_lua
==46720==
Test hello_world...                             [2021/12/07 13:31:48] [ info] [engine] started (pid=46720)
[2021/12/07 13:31:48] [ info] [storage] version=1.1.5, initializing...
[2021/12/07 13:31:48] [ info] [storage] in-memory
[2021/12/07 13:31:48] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2021/12/07 13:31:48] [ info] [cmetrics] version=0.2.2
[2021/12/07 13:31:48] [ info] [sp] stream processor started
[2021/12/07 13:31:48] [ warn] [engine] service will shutdown in max 1 seconds
[2021/12/07 13:31:49] [ info] [engine] service has stopped (0 pending tasks)
[ OK ]
Test append_tag...                              [2021/12/07 13:31:49] [ info] [engine] started (pid=46720)
[2021/12/07 13:31:49] [ info] [storage] version=1.1.5, initializing...
[2021/12/07 13:31:49] [ info] [storage] in-memory
[2021/12/07 13:31:49] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2021/12/07 13:31:49] [ info] [cmetrics] version=0.2.2
[2021/12/07 13:31:49] [ info] [sp] stream processor started
[2021/12/07 13:31:50] [ warn] [engine] service will shutdown in max 1 seconds
[2021/12/07 13:31:50] [ info] [engine] service has stopped (0 pending tasks)
[ OK ]
Test type_int_key...                            [2021/12/07 13:31:50] [ info] [engine] started (pid=46720)
[2021/12/07 13:31:50] [ info] [storage] version=1.1.5, initializing...
[2021/12/07 13:31:50] [ info] [storage] in-memory
[2021/12/07 13:31:50] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2021/12/07 13:31:50] [ info] [cmetrics] version=0.2.2
[2021/12/07 13:31:51] [ info] [sp] stream processor started
[2021/12/07 13:31:52] [ warn] [engine] service will shutdown in max 1 seconds
[2021/12/07 13:31:52] [ info] [engine] service has stopped (0 pending tasks)
[ OK ]
Test type_int_key_multi...                      [2021/12/07 13:31:52] [ info] [engine] started (pid=46720)
[2021/12/07 13:31:52] [ info] [storage] version=1.1.5, initializing...
[2021/12/07 13:31:52] [ info] [storage] in-memory
[2021/12/07 13:31:52] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2021/12/07 13:31:52] [ info] [cmetrics] version=0.2.2
[2021/12/07 13:31:53] [ info] [sp] stream processor started
[2021/12/07 13:31:54] [ warn] [engine] service will shutdown in max 1 seconds
[2021/12/07 13:31:54] [ info] [engine] service has stopped (0 pending tasks)
[ OK ]
Test type_array_key...                          [2021/12/07 13:31:54] [ info] [engine] started (pid=46720)
[2021/12/07 13:31:54] [ info] [storage] version=1.1.5, initializing...
[2021/12/07 13:31:54] [ info] [storage] in-memory
[2021/12/07 13:31:54] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2021/12/07 13:31:54] [ info] [cmetrics] version=0.2.2
[2021/12/07 13:31:55] [ info] [sp] stream processor started
[2021/12/07 13:31:56] [ warn] [engine] service will shutdown in max 1 seconds
[2021/12/07 13:31:56] [ info] [engine] service has stopped (0 pending tasks)
[ OK ]
Test array_contains_null...                     [2021/12/07 13:31:56] [ info] [engine] started (pid=46720)
[2021/12/07 13:31:56] [ info] [storage] version=1.1.5, initializing...
[2021/12/07 13:31:56] [ info] [storage] in-memory
[2021/12/07 13:31:56] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2021/12/07 13:31:56] [ info] [cmetrics] version=0.2.2
[2021/12/07 13:31:57] [ info] [sp] stream processor started
[2021/12/07 13:31:58] [ warn] [engine] service will shutdown in max 1 seconds
[2021/12/07 13:31:58] [ info] [engine] service has stopped (0 pending tasks)
[ OK ]
SUCCESS: All unit tests have passed.
==46720==
==46720== HEAP SUMMARY:
==46720==     in use at exit: 0 bytes in 0 blocks
==46720==   total heap usage: 6,241 allocs, 6,241 frees, 3,483,615 bytes allocated
==46720==
==46720== All heap blocks were freed -- no leaks are possible
==46720==
==46720== For lists of detected and suppressed errors, rerun with: -s
==46720== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

**Comment by tarruda** [comment]
Valgrind logs for filtering with a test script:

```
$ valgrind -s --leak-check=full ./build/bin/fluent-bit -i dummy -F lua -p script=test.lua -p call=cb_filter -m '*' -o stdout
==49734== Memcheck, a memory error detector
==49734== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==49734== Using Valgrind-3.16.1 and LibVEX; rerun with -h for copyright info
==49734== Command: ./build/bin/fluent-bit -i dummy -F lua -p script=test.lua -p call=cb_filter -m * -o stdout
==49734==
Fluent Bit v1.9.0
* Copyright (C) 2019-2021 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2021/12/07 13:53:44] [ info] [engine] started (pid=49734)
[2021/12/07 13:53:44] [ info] [storage] version=1.1.5, initializing...
[2021/12/07 13:53:44] [ info] [storage] in-memory
[2021/12/07 13:53:44] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2021/12/07 13:53:44] [ info] [cmetrics] version=0.2.2
[2021/12/07 13:53:44] [ info] [sp] stream processor started
[0] dummy.0: [1.000000000, {"message"=>"dummy filtered"}]
[1] dummy.0: [1.000000000, {"message"=>"dummy filtered"}]
[2] dummy.0: [1.000000000, {"message"=>"dummy filtered"}]
[3] dummy.0: [1.000000000, {"message"=>"dummy filtered"}]
[0] dummy.0: [1.000000000, {"message"=>"dummy filtered"}]
[1] dummy.0: [1.000000000, {"message"=>"dummy filtered"}]
[2] dummy.0: [1.000000000, {"message"=>"dummy filtered"}]
[3] dummy.0: [1.000000000, {"message"=>"dummy filtered"}]
[4] dummy.0: [1.000000000, {"message"=>"dummy filtered"}]
^C[2021/12/07 13:53:58] [engine] caught signal (SIGINT)
[0] dummy.0: [1.000000000, {"message"=>"dummy filtered"}]
[1] dummy.0: [1.000000000, {"message"=>"dummy filtered"}]
[2] dummy.0: [1.000000000, {"message"=>"dummy filtered"}]
[3] dummy.0: [1.000000000, {"message"=>"dummy filtered"}]
[4] dummy.0: [1.000000000, {"message"=>"dummy filtered"}]
[2021/12/07 13:53:58] [ warn] [engine] service will shutdown in max 5 seconds
[2021/12/07 13:53:58] [ info] [engine] service has stopped (0 pending tasks)
==49734==
==49734== HEAP SUMMARY:
==49734==     in use at exit: 0 bytes in 0 blocks
==49734==   total heap usage: 1,309 allocs, 1,309 frees, 2,360,366 bytes allocated
==49734==
==49734== All heap blocks were freed -- no leaks are possible
==49734==
==49734== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

```lua
function cb_filter(tag, timestamp, record)
   return 1, 1, { message = record.message..' filtered' }
end
```

File: src/flb_lua_messagepack.c [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+    }
+}
+
+/*
+ * This function is to call lua function table.maxn.
+ * CAUTION: table.maxn is removed from Lua 5.2.
+ * If we update luajit which is based Lua 5.2+,
+ * this function should be removed.
+*/
+static int lua_isinteger(lua_State *L, int index);
```
Comment (edsiper) [inline src/flb_lua_messagepack.c:99]: declare this on top if needed.

**Review by edsiper — CHANGES_REQUESTED** [review]

**Comment by edsiper** [comment]
awesome! thanks for this work. 

A couple of tweaks are needed:

- let's rename the new interface from flb_lua_messagepack to just flb_lua
- add a unit test for the new flb_lua interface

**Comment by edsiper** [comment]
the PR title must be changed to something like:

- lua: ....

**Comment by tarruda** [comment]
> awesome! thanks for this work.
> 
> A couple of tweaks are needed:
> 
> * let's rename the new interface from flb_lua_messagepack to just flb_lua
> * add a unit test for the new flb_lua interface

Note that there are two new modules: 

- `flb_lua_messagepack`, which contains functions related to lua/messagepack conversion
- `flb_lua_utils`, which contains utility functions not related to messagepack

Should I merge both modules into `flb_lua` ?

**Comment by tarruda** [comment]
Started a test file for lua, but still working on it

**Comment by tarruda** [comment]
Valgrind logs for new unit test module:

```
$ valgrind ./build/bin/flb-it-lua
==59961== Memcheck, a memory error detector
==59961== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==59961== Using Valgrind-3.16.1 and LibVEX; rerun with -h for copyright info
==59961== Command: ./build/bin/flb-it-lua
==59961==
Test lua_is_valid_func...                       [ OK ]
Test lua_pushtimetable...                       [ OK ]
Test lua_pushmsgpack...                         [ OK ]
Test lua_tomsgpack...                           [ OK ]
SUCCESS: All unit tests have passed.
==59961==
==59961== HEAP SUMMARY:
==59961==     in use at exit: 0 bytes in 0 blocks
==59961==   total heap usage: 8 allocs, 8 frees, 33,984 bytes allocated
==59961==
==59961== All heap blocks were freed -- no leaks are possible
==59961==
==59961== For lists of detected and suppressed errors, rerun with: -s
==59961== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

**Comment by edsiper** [comment]
Looks like there are some linking problems, refs:

https://github.com/fluent/fluent-bit/runs/4465363067?check_suite_focus=true#step:4:1957

**Comment by edsiper** [comment]
thank you! we are almost there, let's fix the following commits messages:

from: 

- lua: Add test module for lua util functions 
- fixup! lua: Add test module for lua util functions 

to 

- tests: internal: lua: add test module for lua util functions 
- lua: Add test module for lua util functions

**Comment by tarruda** [comment]
@edsiper The last commit was just a fixup, which I already squashed. Should be ready to merge after the tests.

**Comment by edsiper** [comment]
thanks!

**Pull request merged**